elements/userlevel/todpdkdevice.cc

```diff
@@ -124,7 +124,16 @@ inline struct rte_mbuf* get_mbuf(Packet* p, bool create=true) {
 
     if (likely(DPDKDevice::is_dpdk_packet(p))) {
         mbuf = (struct rte_mbuf *) p->destructor_argument();
-        p->set_buffer_destructor(DPDKDevice::fake_free_pkt);
+        rte_pktmbuf_pkt_len(mbuf) = p->length();
+        rte_pktmbuf_data_len(mbuf) = p->length();
+        if (p->shared()) {
+            /*Prevent DPDK from freeing the buffer. When all shared packet
+             * are freed, DPDKDevice::free_pkt will effectively destroy it.*/
+            rte_mbuf_refcnt_update(mbuf, 1);
```

Comment:
p->reset_buffer() will be called only if the packet is not shared. If it's not shared, the destructor will indeed never be called and the Packet object will be recycled directly after the call to packet->kill() later. As the refcnt has never been changed by Click, it's still 1 and when DPDK will have sent the Packet, it will be released by DPDK itself.

If it's shared, _destructor is preserved. we increment the refcnt, and when all packet are killed, the _destructor will be called as part of releasing the last ->data_packet().